File: cpp/prob007.cpp
#include <iostream>
#include <chrono>
#include "euler.h"

#define PROB_NUM    7

int compute()
{
    int n = 1, idx = 1;
    while(idx < 10001){
        n += 2;
        if(is_prime(n))
            idx++;
    }
    return n;
}


int test_answer(std::string prob_answer) {
    std::string answer;
    if(!get_answer(PROB_NUM, &answer)) {
        std::cout << "file not found" << std::endl;
        return 1;
    }

    assert(answer.compare(prob_answer) == 0);
    return 0;
}

int main(int argc, char const *argv[])
{
    auto t1 = std::chrono::high_resolution_clock::now();
    int64_t result = compute();
    auto t2 = std::chrono::high_resolution_clock::now();

    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1).count();
    std::cout << result << "\tin " << duration << " us\n";

    // do testing
    test_answer(std::to_string(result));
    return 0;
}

File: rust/src/bin/prob007.rs
extern crate euler;

fn compute() -> u32{
    let mut idx = 1;
    let bound = 10001;

    for n in 2.. {
        if euler::is_prime(n) {
            idx += 1;
            if idx == bound {
                return n;
            }
        }
    }

    return 0;
}

fn main() {
    let now = std::time::Instant::now();
    let result = compute();
    println!("{:?}\tin {}ms", result, now.elapsed().as_millis());
}

#[cfg(test)]
mod tests {
    use super::*;
    const PROB_NUM: usize = 7;
    
    #[test]
    fn test_prob007() {
        let result = compute().to_string();
        euler::test_answer(PROB_NUM, result);
    }
}